cdc: use a new internal table for exchanging generations

kbr- opened the pull request:
Currently when a node wants to create and broadcast a new CDC generation
it performs the following steps:
1. choose the generation's stream IDs and mapping (how this is done is
   irrelevant for the current discussion)
2. choose the generation's timestamp by taking the current time
   (according to its local clock) and adding 2 * ring_delay
3. insert the generation's data (mapping and stream IDs) into
   system_distributed.cdc_generation_descriptions, using the
   generation's timestamp as the partition key (we call this table
   the "old internal table" below)
4. insert the generation's timestamp into the "CDC_STREAMS_TIMESTAMP"
   application state.

The timestamp spreads epidemically through the gossip protocol. When
nodes see the timestamp, they retrieve the generation data from the
old internal table.

Unfortunately, due to the schema of the old internal table, where
the entire generation data is stored in a single cell, step 3 may fail for
sufficiently large generations (there is a size threshold for which step
3 will always fail - retrying the operation won't help). Also the old
internal table lies in the system_distributed keyspace that uses
SimpleStrategy with replication factor 3, which is also problematic; for
example, when nodes restart, they must reach at least 2 out of these 3
specific replicas in order to retrieve the current generation (we write
and read the generation data with QUORUM, unless we're a single-node
cluster, where we use ONE). Until this happens, a restarting
node can't coordinate writes to CDC-enabled tables. It would be better
if the node could access the last known generation locally.

The commit introduces a new table for broadcasting generation data with
the following properties:
-  it uses a better schema that stores the data in multiple rows, each
   of manageable size
-  it resides in a new keyspace that uses EverywhereStrategy so the
   data will be written to every node in the cluster that has a token in
   the token ring
-  the data will be written using CL=ALL and read using CL=ONE; thanks
   to this, restarting node won't have to communicate with other nodes
   to retrieve the data of the last known generation. Note that writing
   with CL=ALL does not reduce availability: creating a new generation
   *requires* all nodes to be available anyway, because they must learn
   about the generation before their clocks go past the generation's
   timestamp; if they don't, partitions won't be mapped to stream IDs
   consistently across the cluster
-  the partition key is no longer the generation's timestamp. Because it
   was that way in the old internal table, it forced the algorithm to
   choose the timestamp *before* the generation data was inserted into
   the table. What if the inserting took a long time? It increased the
   chance that nodes would learn about the generation too late (after
   their clocks moved past its timestamp). With the new schema we will
   first insert the generation data using a randomly generated UUID as
   the partition key, *then* choose the timestamp, then gossip both the
   timestamp and the UUID.
   Observe that after a node learns about a generation broadcasted using
   this new method through gossip it will retrieve its data very quickly
   since it's one of the replicas and it can use CL=ONE as it was
   written using CL=ALL.

The generation's timestamp and the UUID mentioned in the last point form
a "generation identifier" for this new generation. For passing these new
identifiers around, we introduce the cdc::generation_id_v2 type.

Fixes #7961.

---

I have separated the first 12 commits, which are refactors and improvements to the existing algorithm, into a separate PR: https://github.com/scylladb/scylla/pull/8283.

I need to do some more testing, but review can be conducted in the meantime.

For optimal review experience it is best to first read the updated design notes (you can read them rendered here: https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md), specifically the ["Generation switching"](https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md#generation-switching) section followed by the ["Internal generation descriptions table V1 and upgrade procedure"](https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md#internal-generation-descriptions-table-v1-and-upgrade-procedure) section, then read the commits in topological order.

kbr- commented:
Updated the PR with some bugfixes, improvements, removed TODOs. Not listing the changes since no review was done yet.

kbr- commented:
I updated the design notes. For optimal review experience it is best to first read the notes (you can read them rendered here: https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md), specifically the ["Generation switching"](https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md#generation-switching) section followed by the ["Internal generation descriptions table V1 and upgrade procedure"](https://github.com/kbr-/scylla/blob/cdc-gen-table/docs/design-notes/cdc.md#internal-generation-descriptions-table-v1-and-upgrade-procedure) section, then read the commits in topological order.

I also created a PR to update CDC dtests (https://github.com/scylladb/scylla-dtest/pull/2035). The updated dtests have passed on a dev build.

kbr- commented:
I've limited the size of mutations used when inserting the CDC generation data to the new table. The size is now inversely proportional to the concurrency and number of replicas. Diff: https://github.com/scylladb/scylla/compare/81acd6badcc7bceb46c0c1572a2068195c7d92da..3c046b3db01ab4491e0ace18e35bbeb01d791e70
the comment seen in the diff explains my reasoning.

With this all TODOs seem to be resolved, but I'm still testing the change.

kbr- commented:
@haaawk @avikivity @elcallio ping: any chance for getting a review soon?

kbr- commented:
Removed RFC: prefix since it's basically a finished PR.

elcallio commented on cdc/generation.cc:
```diff
@@ -928,17 +928,17 @@ future<bool> generation_service::do_handle_cdc_generation(db_clock::time_point t
+    // We assume that generation changes don't happen ``too often'' so every node can learn about a generation
+    // before it is superseded by a newer one which causes nodes to start gossiping the about the newer one.
+    // The assumption follows from the requirement of bootstrapping nodes sequentially.
+    if (!_gen_id || *_gen_id < ts) {
```
Is sequential bootstrap really enough to guarantee something like that? Even if nodes are started sequentially, I assume negotiating/disseminating cdc generation info etc takes at least some time. Could not a new node get going during this time, creating an overlap? Is this an issue?
What about nodes crashing & coming back up / decommission-recommission? I assume the distributed mechanism block somehow to finish the current iteration before getting caught up in new things?

kbr- replied:
The system does not prevent the admin from (all) mistakes such as bootstrapping nodes non-sequentially - this is why we want to switch to "Raft-based topology changes", where all decisions about topology changes will be serialized. So yes, it may be possible that the admin does a mistake and tries to join two nodes at the same time, causing two generations to be broadcasted at the same time, causing the one with the smaller timestamp to be learned by only a subset of the nodes (and that would be bad).

There is some small protection: when a node joins it first performs a gossip round and checks the STATUS of other nodes. If it has status=BOOT and sees another node with status=BOOT it will refuse to join (this check happens before we introduce a new generation). But of course races are possible there where the protection fails, it's not perfect. For now we mostly hope that admins follow documentation (which explicitly says to wait between bootstrapping nodes). If they don't, disappearing CDC generations are not their only problem (possible data loss is another).

kbr- replied:
If nodes are bootstrapped sequentially then we're good because the CDC generation broadcast should finish before even streaming starts (we wait for ring_delay after introducing a new generation and before starting streaming data). The next node joins after streaming finishes for this node.

Nodes crashing & restarting may be a problem but it would have to be a coordinated crash. Something like:
- node X creates and starts gossiping a new generation and then immediately crashes
- a minute later it restarts, continues gossiping it, but it means that other nodes in the cluster learn about the generation too late.

This scenario can't happen on bootstrap because if node X was bootstrapping when it crashed then on restart it will create a completely new generation, so it's as if we never created the previous one.
But it can potentially happen on checkAndRepairCdcStreams. It is highly improbable, the node would have to crash precisely after it inserts the generation ID into application state but before it reaches some other node. Still, perhaps it would be good to somehow prevent this, e.g. persisting the timestamp only after it reaches at least one other node (so if we crash in this scenario, we forget about the generation). But this can also break.
There is a solution that preserves consistency under any crash scenarios and is somewhat available but it requires consensus, hence we wait for Raft before implementing it.

elcallio commented on service/storage_service.cc:
```diff
@@ -586,33 +586,46 @@ void storage_service::join_token_ring(int delay) {
+                        "Attempted to create a new CDC generation on node start since it was missing, but failed"
+                        " due to a partially upgraded cluster (the CDC_GENERATIONS_V2 feature is not enabled)."
+                        " If you're upgrading from Scylla <= 4.2 to Scylla >= 4.5, rollback and do a proper upgrade (single"
+                        " major version at a time). Otherwise finish the upgrade and then restart this node again.");
```
Is not upgrading between 4.x:s _minor_ version upgrades? 4 -> 5 == major

kbr- replied:
Ok, I think you're right. But how is 4.x.y -> 4.x.(y+1) called?

elcallio commented:
Fwiw, looks ok to me. Gargatuan PR, which imho is a bit of a negative, and certainly does not help reviewability. Added a few conceptual questions that I am sure the author already dealt with, but I would like to know more about (because I am so curious).

kbr- commented:
@elcallio it will be much better if we merge https://github.com/scylladb/scylla/pull/8283 first - which is a prefix of this PR that contains mostly refactors, no "controversial" changes so should be easy to review.
Then this PR will only have a couple of commits and won't have so many tree-wide changes.

haaawk commented on cdc/generation.cc:
```diff
@@ -330,7 +330,7 @@ topology_description limit_number_of_streams_if_needed(topology_description&& de
     return topology_description(std::move(entries));
 }
 
-future<cdc::generation_id> make_new_cdc_generation(
```
rename generation_id -> generation_id_v1 should be a separate commit!!!

kbr- replied:
That's not logically a rename, some places still use `generation_id` (which is now a variant), this place stops using `generation_id` and starts using `generation_id_v1`. The fact that `generation_id_v1` is the same type as `generation_id` was before this commit is a "coincidence".

If I did a commit that renames `generation_id` to `generation_id_v1` before this commit then in this commit I would have to rename some instances `generation_id_v1` back to `generation_id`. Hard and unnecessary work.

haaawk commented on cdc/generation.cc:
```diff
@@ -975,15 +975,27 @@ shared_ptr<db::system_distributed_keyspace> generation_service::get_sys_dist_ks(
 db_clock::time_point get_ts(const generation_id& gen_id) {
-    return gen_id.ts;
+    return std::visit(make_visitor(
+    [] (const generation_id_v1& id) { return id.ts; },
+    [] (const generation_id_v2& id) { return id.ts; }
```
I'm wondering whether @avikivity would agree to make this a generic lambda :)

haaawk commented on gms/versioned_value.cc:
```diff
@@ -99,7 +110,37 @@ std::optional<cdc::generation_id> versioned_value::cdc_generation_id_from_string
+        constexpr auto invalid_format_template = "Invalid value of CDC generation ID string: {}. The format is \"v2;<timestamp>;<uuid>\".";
+        const char* const end = s.c_str() + s.size();
+
+        long long ts;
```
int64_t?

haaawk commented on cdc/generation.cc:
```diff
@@ -330,7 +330,7 @@ topology_description limit_number_of_streams_if_needed(topology_description&& de
 }
 
-future<cdc::generation_id> make_new_cdc_generation(
+future<cdc::generation_id_v1> make_new_cdc_generation(
```
Comment to commit msg:
```
 and how to persist new identifiers in local tables.
```
Is this really the case?

kbr- replied:
Yes, `update_cdc_generation_id` and `get_cdc_generation_id`, which update and retrieve the ID from the local table, has been modified in this commit to handle both types of identifiers.

So nodes can handle the new IDs after this commit but they don't create these IDs yet (in this commit).

haaawk commented on db/system_distributed_keyspace.cc:
```diff
@@ -90,6 +90,37 @@ schema_ptr cdc_generations() {
+                /* The value of the `ignore_msb` sharding parameter of the node which was the owner of this token range
+                 * when the generation was first created. Together with the set of streams above it fully describes
+                 * the mapping for this particular range. */
+                .with_column("ignore_msb", byte_type)
```
This solves the problem of a node being replaced with a machine that has different number of CPUs, right? Can we retire IGNORE_MSB from gossip now?

haaawk replied:
NVM. It does not solve the CPU problem but maybe we could add a column with number of shards and that way fix the CPU problem?

haaawk replied:
Actually length of streams is the number of CPUs so all's good.

kbr- replied:
`IGNORE_MSB` in gossip is used when creating new CDC generations so we can't remove it (unless we introduce a new method of obtaining this parameter).

haaawk commented on db/system_distributed_keyspace.cc:
```diff
@@ -246,6 +287,12 @@ future<> system_distributed_keyspace::remove_view(sstring ks_name, sstring view_
  * possible replica. Until another node joins, reads will also use CL = ONE, retrieving the data
  * from the only existing replica.
+ *
+ * With system_distributed_v2 tables things are simpler since they are using the Everywhere
```
Maybe name the keyspace somehow more meaningful? For example, system_distributed_everywhere?

avikivity commented:
Ugh, I thought this was already merged. I'll review it soon.

avikivity commented:
Wait, how does this relate to 67d4e5576d37aac2c5add4fbbc3b2e351ad0836d? Is it a different table?

avikivity commented:
@gleb-cloudius @kostja @tgrabiec such tables are prime candidates for raft group 0. I don't propose mixing this with raft now, but this is one of the services that raft should provide sooner - everywhere replication strategy tables. It should be easy(ier) on one hand, and provide good value on the other.

kbr- commented:
> Wait, how does this relate to 67d4e55? Is it a different table?

Yes, 67d4e55 replaced the client table, this one replaces the internal table.
It's the successor to https://github.com/scylladb/scylla/pull/8169.

avikivity commented on cdc/generation.cc:
```diff
@@ -330,7 +330,7 @@ topology_description limit_number_of_streams_if_needed(topology_description&& de
     return topology_description(std::move(entries));
 }
 
```
Very good, types that refer to the role of the value are better than types that just describe the structure.

avikivity commented on cdc/generation_id.hh:
```diff
@@ -28,10 +28,17 @@
+};
+
+using generation_id = std::variant<generation_id_v1, generation_id_v2>;
+
```
Can we extend rather than add a variant? I think it will simplify many paths.

If we see a v1 generation id, append to it a fixed uuid (perhaps a null uuid, perhaps just some uuid we will generate once and hardcode). This will unify many code paths.

avikivity replied:
For example, operator== for generation ids can be auto-generated with `operator<=>(const generation_id&) = default`.

kbr- replied:
Having two separate types for v1 and v2 generation IDs protects us (in compile time) from introducing v2 generation IDs to the cluster by accident - for example, it would be very undesirable to introduce a v2 ID to a cluster in which some nodes do not yet understand the new format.

I want these different codepaths to be different, and I want the type system to force them to be different.

avikivity replied:
How does it protect? We can still send a v1 where a v2 is expected.

kbr- replied:
Functions which expect a v1 cannot get a v2 accidentally (`rewrite_streams_descriptions`, `cdc_set_rewritten`) because the types are different (and neither is a subtype of the other).

I also know where a v2 is first introduced: `make_new_cdc_generation` is the only function which returns a v2 and I can protect all places where this function is used with a feature check.

Which code paths do you think would be unified?

avikivity replied:
Ok, I see. Hypothetically, if you introduced a new gossip application state for v2 you wouldn't need the variant? That is, the variant multiplexes the two types for transport on gossip and then de-multiplexes them on arrival?

kbr- replied:
> Hypothetically, if you introduced a new gossip application state for v2 you wouldn't need the variant?

I don't strictly _need_ the variant. I could make v1 a special case of v2, but then I would have to be more careful not to accidentally send a v1 as a v2, for example. I just think that variant is an elegant method of protecting us from making a mistake.

Introducing a new gossip application state wouldn't change much in that regard. It could possibly allow us to perform upgrade without a cluster feature (because the existing gossip app state would always have v1 IDs and the new gossip app state would always have v2 IDs; non upgraded nodes would keep using existing gossip app state without risk of trying to read a v2 ID).

> That is, the variant multiplexes the two types for transport on gossip and then de-multiplexes them on arrival?

Yes. I could do it without a variant, for example:
```
struct generation_id {
    db_clock::time_point ts;
    std::optional<utils::UUID> uuid;
}
```
then when I insert the ID into gossip, I serialize it as v2 if and only if `uuid` is not nullopt (and otherwise I serialize it as v1).
Could also do without `optional` by using a special hardcoded `utils::UUID` but then I would have to distinguish v1 from v2 by comparing `uuid` field with this hardcoded value - if it's equal, serialize as v1, otherwise as v2.

kbr- replied:
Even if I introduced a new gossip application state, I would still have to be careful not to send a v2 through the old appstate or a v1 through the new. The variant would give the same protection here as it does when we use a single appstate.

avikivity commented on database.cc:
```diff
@@ -1136,7 +1136,8 @@ void keyspace::mark_as_populated() {
 
 
```
> -  the data will be written using CL=ALL and read using CL=ONE; thanks
>    to this, restarting node won't have to communicate with other nodes
>    to retrieve the data of the last known generation. Note that writing
>    with CL=ALL does not reduce availability: creating a new generation
>    *requires* all nodes to be available anyway, because they must learn
>    about the generation before their clocks go past the generation's
>    timestamp; if they don't, partitions won't be mapped to stream IDs
>    consistently across the cluster

This means that if a node is down, we cannot perform a node addition or removal.

It was already iffy, since we don't yet enable @asias repair-based node operations (which do allow node additions and removal with nodes down).

This is a negative side effect of this series, but I think we can tolerate it because those operations were quite shaky before. Once we have raft both node operations and everywhere replication will be safer in the face of downed (or partitioned) nodes.

kbr- replied:
> repair-based node operations (which do allow node additions and removal with nodes down).

Do they? Decommissioning a node when some nodes are down may cause data loss. Today decommissioning ensures that all data is moved to other nodes. Consider RF=1 table - you must move data from the leaving node to _every other node_ (assuming that every node owns at least one token range), so you cannot allow some nodes to be down during the operation. There is also `removenode` which is the unsafe counterpart of decommission (one which allows dropping data). And our docs explicitly warn from using `removenode`. (https://docs.scylladb.com/operating-scylla/procedures/cluster-management/remove_node/)

Similarly node bootstrap requires all nodes to be up (if we don't want data loss) because data must be streamed from _every_ node. Again consider RF=1 case.

CDC always assumed that all nodes are up when bootstrapping a node, the assumption does not change with this PR.

avikivity replied:
Obviously RF=1 cannot be decommissioned with a node down (@asias how will repair-based-node-ops react?). But with RF=3, it will either move the data to the replica that will own the data (if it is up) or to two other replicas (if not), preserving quorum.

(there is a negative effect to this in that a large repair will be needed later).

kbr- replied:
Even with RF=3 you could have a successful CL=ONE write which ended up on exactly one replica, and is supposed to be observable by all later successfull CL=ALL reads unless a data-loss event happens in between which forces the admin to perform an unsafe administrative operation such as `nodetool removenode` (which - as the documentation properly points out - does not preserve consistency of operations). `nodetool decommission` is supposed to be a **safe** operation, preserving consistency, in particular CL=ONE writes -> CL=ALL reads consistency ("read-your-writes"). If the admin accepts possibility of dropping consistency and really wants to remove a node when some nodes are down, they are supposed to use `nodetool removenode`.

Bootstrap is symmetric to `nodetool decommission` and is also supposed to be a completely safe operation, preserving consistency of operations and not losing data. For that it **must** require all other nodes to be up in general.

I see no reason why we would want to lower the requirements for bootstrap. If some nodes are irrecoverably down, the user should replace or remove them and bring a cluster to a total UN state before bootstrapping a new node.

avikivity replied:
CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.

The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.

kbr- replied:
> CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.

I'm considering a scenario where no replica is lost. Should we drop our promises in that case as well?

The very true generalization of your statement is: for any RF, CL=(whatever) write + CL=(whatever) read is not a safe operation, because you could lose RF replicas before you replace any of them (doesn't matter if you repair or not).

That's right, but **if no disaster happens**, then we should guarantee that CL=QUORUM reads observe CL=QUORUM writes. Similarly, we should guarantee that CL=ALL reads observe CL=ONE writes.

In fact it's common knowledge in the Cassandra community - and I'm guessing Scylla community too, since people assume they give similar guarantees - that in order to read back previously written data, the following must be satisfied: `write CL + read CL > RF`. For example: https://stackoverflow.com/questions/25043208/read-operation-in-cassandra-at-consistency-level-of-quorum. As the accepted answer says, examples are: `write ALL + read ONE`, `write ONE + read ALL`, `write QUORUM + read QUORUM`. When I first heard about Scylla on a lecture by @haaawk at my university, that's what was written on the presentation too.

> The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.

Ok, but bootstrap should be safe (not allow consistency loss) by default. Bootstrapping in presence of downed nodes should be an override config option and heavily discouraged in docs (reserved only for disaster scenarios). The override should contain a promise from the administration that the downed node is lost permanently. Given such a promise we can also safely introduce a new CDC generation. The CL=ALL write in this table would be then changed to CL=ALL-ONE write.

avikivity replied:
> > CL=ONE write + CL=ALL read is not a safe operation. The replica could be lost before a repair happens.
> 
> I'm considering a scenario where no replica is lost. Should we drop our promises in that case as well?
> 

We should see if those promises are worth keeping. CL=ALL means that as soon as you perform maintenance of have a crash you lose availability. Write(CL=ONE) + read(CL=ALL) gives really good availability for write, but poor durability. More reasonable is write(CL=ALL) + read(CL=ONE). Compromising maintainability for these uncommon scenarios is not a good tradeoff IMO.

> The very true generalization of your statement is: for any RF, CL=(whatever) write + CL=(whatever) read is not a safe operation, because you could lose RF replicas before you replace any of them (doesn't matter if you repair or not).

RF=3 only handles one node failure. You must handle this node failure soon; until you do, another failure can cause you data loss. If you want to survive two node failures, you must use RF=5.

> 
> That's right, but **if no disaster happens**, then we should guarantee that CL=QUORUM reads observe CL=QUORUM writes. Similarly, we should guarantee that CL=ALL reads observe CL=ONE writes.

QUORUM/QUORUM safety is maintained with repair-based node operations. Write(ONE)/Read(ALL) might not be, if the node is downed before it had a chance to replicate, but write(ONE) implicitly gives up durability because a disaster *might* happen. Write(ALL)/Read(ONE) is still safe.

> 
> In fact it's common knowledge in the Cassandra community - and I'm guessing Scylla community too, since people assume they give similar guarantees - that in order to read back previously written data, the following must be satisfied: `write CL + read CL > RF`. For example: https://stackoverflow.com/questions/25043208/read-operation-in-cassandra-at-consistency-level-of-quorum. As the accepted answer says, examples are: `write ALL + read ONE`, `write ONE + read ALL`, `write QUORUM + read QUORUM`. When I first heard about Scylla on a lecture by @haaawk at my university, that's what was written on the presentation too.

Well, if permanent node failures are not ruled out then write(ONE)+read(ALL) will not work, either in Cassandra or Scylla.

> 
> > The reason to allow bootstrap in the face of downed nodes is orthogonality. It may be unfeasible to resurrect a node in time, addition the dependency between fixing the node and adding new nodes means that scaling up and down is not determined by the streaming bandwidth, but by the sum of the time to fix a node and streaming bandwidth.
> 
> Ok, but bootstrap should be safe (not allow consistency loss) by default. Bootstrapping in presence of downed nodes should be an override config option and heavily discouraged in docs (reserved only for disaster scenarios). The override should contain a promise from the administration that the downed node is lost permanently. Given such a promise we can also safely introduce a new CDC generation. The CL=ALL write in this table would be then changed to CL=ALL-ONE write.

It is too strict. The downed node may be recovered, and repair-based node operations will ensure data safety. W1RA is too rare to align everything with its requirements.

For CDC, we should use Raft instead of these heuristics.

kbr- replied:
> We should see if those promises are worth keeping. CL=ALL means that as soon as you perform maintenance of have a crash you lose availability. Write(CL=ONE) + read(CL=ALL) gives really good availability for write, but poor durability. More reasonable is write(CL=ALL) + read(CL=ONE). Compromising maintainability for these uncommon scenarios is not a good tradeoff IMO.

W+R>RF is a simple to remember and understand rule, and users expect it to work intuitively; e.g. if you wrote data to one replica and read from all, you contact that one replica, so you see the data. Introducing exceptions to this rule makes systems built on top of Scylla/Cassandra harder to reason about and understand.

Another path we could take while keeping simple rules is to drop all CLs other than QUORUM altogether, high availability and consistency, and tolerance of RF/2 failures. AFAIR DynamoDB took this approach.

> RF=3 only handles one node failure.

I think you implicitly assume quorum operations. Under some other strategies it can handle 2 failures. For example consider the following strategy. Users always read with CL=1, and:
- until 3 nodes are available, write with CL=3
- after one failure, write with CL=2
- after second failure, write with CL=1

With this strategy we remain consistent and available and tolerate up to 2 failures.

> but write(ONE) implicitly gives up durability because a disaster _might_ happen

"write(TWO) implicitly gives up durability because 2 disasters _might_ happen" (correlated failure).

I don't think it's a question of what _might_ happen. Write(ONE) tolerates 0 failures, yes. But that doesn't mean Scylla should be the one to actually introduce the failure. Allowing bootstrap with a partitioned or temporarily shut down node is introducing a failure.

> Well, if permanent node failures are not ruled out then write(ONE)+read(ALL) will not work, either in Cassandra or Scylla.

And what should we guarantee **before** the failure happens?

> It is too strict. The downed node may be recovered, and repair-based node operations will ensure data safety. 

So even if  the down node is recoverable we still want to allow bootstrap to proceed, just so the entire procedure is a bit faster? I don't see much value here, considering the rarity of those operations. You're considering a situation where a node permanently fails and then very soon after the admin decides that it's a good moment to scale up. I think that when nodes permanently fail, admins' first priority is to stabilize the cluster (replacing the failed nodes), not scaling up.

> For CDC, we should use Raft instead of these heuristics.

Raft does not automatically solve every problem. Maybe it would if everything would go through a single state machine replicated by a single Raft group, but such solutions don't scale. Thus you need communication between independent Raft state machines, so you must solve problems "outside Raft", and CDC generation management is full of problems that must be solved outside Raft.

That said, when we introduce **Raft-driven tablets**, then it will be possible to safely introduce a new CDC generation without all nodes being up - only the leader of each tablet will have to be available.

asias replied:
I have been using the "CL=ONE write + CL=ALL read" example to convince Avi that we should reject the bootstrap by default if a node is down during bootstrap instead of stream from quorum nodes and continue. 

In case the node is down temporarily, we could avoid data loss if we reject the bootstrap ops.

In case the node is really down and not way to recover, or there is no enough time to recover the node, or the user does not bother to recover the node and think it is more important to add a new node immediately than losing the data, we can use the ignore_nodes to exclude the down nodes explicitly like we do in the latest removenode and replacing and regular repair operation.

It is much easier for user to reason about the system:  we are on the safe side by default not matter what, if you want to bootstrap  with the node down you might lose data, users have to ask for it explicitly. 

There are other reason we need all nodes to be up during bootstrap. A node might be partitioned from the bootstrap node, it is not down and can server user reads /writes, if we continue the bootstrap, the partitioned node will use wrong topology which can cause data consistency issue.  Again, to be safe by default, we should reject the bootstrap in such case.

asias replied:
> > For CDC, we should use Raft instead of these heuristics.
> 
> Raft does not automatically solve every problem. Maybe it would if everything would go through a single state machine replicated by a single Raft group, but such solutions don't scale. Thus you need communication between independent Raft state machines, so you must solve problems "outside Raft", and CDC generation management is full of problems that must be solved outside Raft.

It is the same for topology change. People think raft will magically allow us to do topology change with some nodes down since we can achieve some quorum using raft. Raft will not help as long as we still use distributed tokens to form the token ring for read / write. We need all nodes to be aware of this change to maintain consistency.

avikivity replied:
With Raft, a node that is partitioned from the rest of the cluster will be aware that it is partitioned and will not break consistency. It can be done this way:

1. Raft creates a "topology version number" each time topology changes
2. A coordinator includes the topology version number in each request
3. Replicas reject requests that have mismatched topology version number

Note that each topology change generates two versions. Let's assume we start from version v0.
1. Node is added, we start double-writes. Generate v1 and tell nodes to accept v1 and v0.
2. All live nodes are aware of v1. Tell them to reject v0 and wait for ack.
3. Start streaming data (maybe via repair).
4. Streaming is complete, generate v2 that has the new topology without double writes (but accept v1).
5. All live nodes are aware of v2, Tell them to reject v1 and execute cleanup.

kbr- replied:
Consider a cluster of 9 nodes and a keyspace with RF = 3. Suppose there are 3 "tablets" / "replica sets" / "Raft groups" (however you want to name them) replicating different parts of the keyspace: {n1, n2, n3}, {n4, n5, n6}, {n7, n8, n9} (there may be other unlisted tablets, e.g. {n1, n4, n7}, but we don't care for this discussion). Suppose that n1 is the leader/coordinator of the first group, n4 of the second, n7 of the third. Suppose that the third group is partitioned from the rest. Recall that all nodes are also members of Raft Group 0 which handles topology. Observe that {n1, n2, n3, n4, n5, n6} form a quorum in Raft Group 0, thus they can reach a decision on changing the topology of the cluster. Unfortunately, even though the topology was changed, the group {n7, n8, n9} does not know about it because it was partitioned. However, the entire group is running and all nodes within the group can contact each other, so the leader of the group - n7 - can still perform writes, unaware of the topology change. It keeps using the same "topology version number".

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -90,6 +90,37 @@ schema_ptr cdc_generations() {
+                .with_version(system_keyspace::generate_schema_version(id))
+                .build();
+    }();
+    return schema;
```
Prepare for a v3 of this table once we have tablets!

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -330,6 +377,122 @@ system_distributed_keyspace::read_cdc_topology_description(
+
+    // It has been tested that sending 1MB batches to 3 replicas with concurrency 20 works OK,
+    // which would correspond to L ~= 60MB. Hence that's the limit we use here.
+    const size_t L = 60 * 1000 * 1000;
```
`60'000'000`

avikivity replied:
It is sad that high-level code has to resort to these calculations. We should eventually move them towards the core - this code should only provide a stream of small mutations, and the core should take care of batching. But we can do this when we have more users (and so know more about the requirements).

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -543,7 +543,7 @@ system_distributed_keyspace::create_cdc_desc(
             { std::move(m) },
             quorum_if_many(ctx.num_token_owners),
-            db::timeout_clock::now() + 10s,
+            db::timeout_clock::now() + 30s,
```
I hope we don't get into a situation where we livelock without the ability to complete the operation. In theory we could increase the delta after a failure and retry, but this would be never-tested code. Let's hope this is sufficient.

In the raft era, topology changes will be atomic with cdc changes (cdc stream routing is just an aspect of topology).

avikivity commented:
Kudos for such an easy to review series. The commits are well motivated and explained to a non-expert.

I reviewed mostly the interaction of the series with the non-cdc elements of the system, and it looks good in that regard. I'll leave the reviewing of the series wrt. cdc itself to people more familiar (I did not see anything problematic).

The only issue I see is that unifying the two generation id types can simplify the code significantly.

kbr- commented:
v2:
- rebase onto new version of #8283
- `long long` -> `int64_t`
- `system_distributed_v2` -> `system_distributed_everywhere`
- `major version` -> `minor version`
- `60 * 1000 * 1000` -> `60'000'000`

Diff: https://github.com/scylladb/scylla/compare/3c046b3db01ab4491e0ace18e35bbeb01d791e70..3b2af17caa0e5b2aa966a37c6af145e1b62898fd

avikivity commented:
It's better not to change the target of the pull request to next, it can show the request as merged even if it is dequeued later.

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -168,6 +200,15 @@ future<> system_distributed_keyspace::start() {
+        auto ksm = keyspace_metadata::new_keyspace(
+                NAME_EVERYWHERE,
+                "org.apache.cassandra.locator.EverywhereStrategy",
+                {},
```
Please split out the new keyspace. This one is important and deserves its own commit (maybe its own pull request). I'm in favor, but we need to consider the impact here.

kbr- replied:
I will create a new PR after #8283 gets merged because I don't want to deal with rebase hell (I think there will be some conflicts).

avikivity commented on cdc/generation.cc:
```diff
@@ -299,38 +299,6 @@ bool should_propose_first_generation(const gms::inet_address& me, const gms::gos
-size_t limit_of_streams_in_topology_description() {
-    // Each stream takes 16B and we don't want to exceed 4MB so we can have
-    // at most 262144 streams but not less than 1 per vnode.
-    return 4 * 1024 * 1024 / 16;
```
What about a mixed cluster? This may have to wait for a while.

kbr- replied:
As the commit message explains:
> New nodes no longer insert generations into the old generations table
(which stored the entire generation in a single cell), they use the new
format instead. It is thus no longer necessary to limit the sizes of
created generations.

A node that contains this commit will not introduce CDC generations in the old format. For the new format it will refuse to introduce a new generation until the cluster is fully upgraded due to the feature flag check.

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -73,23 +73,6 @@ schema_ptr view_build_status() {
-        auto id = generate_legacy_id(system_distributed_keyspace::NAME, system_distributed_keyspace::CDC_TOPOLOGY_DESCRIPTION);
-        return schema_builder(system_distributed_keyspace::NAME, system_distributed_keyspace::CDC_TOPOLOGY_DESCRIPTION, {id})
-                /* The timestamp of this CDC generation. */
-                .with_column("time", timestamp_type, column_kind::partition_key)
```
What about mixed clusters?

kbr- replied:
Removing this code causes the node to no longer attempt to announce the table creation when it starts. It won't cause the table, if it exists, to disappear (this is a normal distributed table, the only special thing about it is that the schema mutations of this table use `min_timestamp`). Thus in mixed clusters the table will still be there. In new clusters there will be no node to announce it so the table won't appear.

avikivity commented on docs/design-notes/cdc.md:
```diff
@@ -247,4 +261,4 @@ Note that every node will perform this procedure on upgrade, but there's a high
 
 #### TODO: expired generations
-The `expired` column in `cdc_generation_timestamps` and `cdc_generation_descriptions` means that this generation was superseded by some new generation and will soon be removed (its table entry will be gone). This functionality is yet to be implemented.
+The `expired` column in `cdc_generation_timestamps` means that this generation was superseded by some new generation and will soon be removed (its table entry will be gone). This functionality is yet to be implemented.
```
Please also add documentation about the new keyspace (in that commit).

avikivity commented:
Looks good. I mainly reviewed from the point of view of integration with the rest of the system, not cdc (which I have limited knowledge of). @nyh please review that.

Two pending items: about the generation_id variant (if I understand it correctly, it's acceptable) and about system_distributed_everywhere, which can't be snuck in like that.

nyh commented:
I'll review now, but please note that: 1. there are conflicts so it needs rebasing, and 2. Avi said above he still has two pending items.

kbr- commented:
@nyh please start with PR #8283 which is less "controversial" and therefore easier to merge (it is a prefix of this PR). I also already rebased it. There are no pending items there.

kbr- commented:
I separated PR #8457 with a single commit which adds the keyspace, rebased this PR on top of this commit, and added a line in cdc.md about the keyspace:
https://github.com/scylladb/scylla/compare/93d06cedcb2d15724e0e90cb88e1fcefbb422aef..7c48d10fed543e8dedb705428dc627e1a4509f26

I'll rebase the PR again on top of master after #8457 gets merged.

kbr- commented:
@avikivity the keyspace is finally in master. I rebased the PR, it is ready to be merged.
Note: for gating to pass, we need to simultaneously merge https://github.com/scylladb/scylla-dtest/pull/2035.

avikivity commented:
This simultaneous merging is impossible to arrange. Can you change the dtest to pass with either the old or new code?

Usually, these cases indicate that we're breaking something. What if the user depends on the same thing dtest does?

kbr- commented:
> This simultaneous merging is impossible to arrange. Can you change the dtest to pass with either the old or new code?

Gating will fail once but then pass the second time when both PRs are merged.
I can modify the dtest but why not take the lazy path... Gating fails constantly, one more failure won't hurt. And we can speed up the process by instantly aborting after the first merge.

> Usually, these cases indicate that we're breaking something. What if the user depends on the same thing dtest does?

The reason dtest fails is because it creates a new cluster on each test, and the new cluster contains only the new internal tables in new versions of Scylla and only old internal tables in old versions of Scylla, and the dtest uses the corresponding tables (so old dtest doesn't work with new Scylla, new dtest doesn't work with old Scylla). If new clusters in new versions of Scylla created and maintained both versions of internal tables then we could merge this PR followed by the dtest PR, but creating and maintaining both internal table versions in new clusters doesn't make any sense.

In any case the user should not depend on undocumented internal tables that are subject to change (they are only internally documented in cdc.md).

avikivity commented:
@kbr- ok. Get the dtest merge and ping me.

kbr- commented:
> @kbr- ok. Get the dtest merge and ping me.

@avikivity ping as requested

kbr- commented:
Thanks!

avikivity commented:
Dequeued, dtest doesn't pass next promotion.

asias commented:
> Dequeued, dtest doesn't pass next promotion.

I saw a lot of problems debugging the next branch containing this patch set.

asias commented:
> > Dequeued, dtest doesn't pass next promotion.
> 
> I saw a lot of problems debugging the next branch containing this patch set.

@kbr- Please run simple tests like start one node then start another. You will see the problems.

kbr- commented:
> @kbr- Please run simple tests like start one node then start another. You will see the problems.

I did run such tests and more complex ones but I forgot to enable RBO. Without RBO everything works.
